include/circt/Dialect/Handshake/HandshakeOps.td

```diff
@@ -1249,3 +1249,131 @@ def EndOp : Handshake_Op<"end", [Terminator]> {
     mlir::ValueRange getMemoryControls();
   }];
 }
+
+//===----------------------------------------------------------------------===//
+// Speculation operations
+//===----------------------------------------------------------------------===//
+
+def SpeculatorOp : Handshake_Op<"speculator", [ 
+  TypesMatchWith<"data operand type matches data result type",
+                 "dataIn", "dataOut", "$_self">]> {
+  let summary = "Central control unit of the speculative circuit.";
+  let description = [{
+    The speculator produces speculative tokens when real data has
+    not yet arrived. The speculator is also responsible for deciding 
+    the correctness of its speculation.
+
+    It needs to send control tokens to save units, commit units, 
+    and save-commit units.
+
+    Example:
+    ```mlir
+    %dataOut, %saveCtrl, %commitCtrl, %SCControl1, %SCControl2, 
+      %SCBranchControl = "handshake.speculator"(%dataOut) 
+      {name = #handshake.name<"speculator0">} : 
+      (i1) -> (i1, i1, i1, i3, i3, i1)
+    ```
+  }];
+
+  let arguments = (ins AnyType : $dataIn);
+  let results = (outs AnyType : $dataOut,
+                  I1 : $saveCtrl, I1 : $commitCtrl, 
+                  I<3> : $SCControl1, I<3> : $SCControl2, 
+                  I1 : $SCBranchControl);
```

Comment:
All operations are missing the `assemblyFormat` field. A default implementation may be generated automatically but it's good practice to have the format explicitly specified so that it's easier for the reader to understand the operation's format. I recommend taking inspiration from other Handshake operations to understand MLIR's expected format for this field (otherwise, refer to the [official documentation](https://mlir.llvm.org/docs/DefiningDialects/Operations/#declarative-assembly-format)).